In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/muhammadumer7804/social-media-dataset/instagram_usage_lifestyle.csv


In [2]:
import pandas as pd

df = pd.read_csv('/kaggle/input/datasets/muhammadumer7804/social-media-dataset/instagram_usage_lifestyle.csv')
print(df.columns.tolist())
df.head()

['user_id', 'app_name', 'age', 'gender', 'country', 'urban_rural', 'income_level', 'employment_status', 'education_level', 'relationship_status', 'has_children', 'exercise_hours_per_week', 'sleep_hours_per_night', 'diet_quality', 'smoking', 'alcohol_frequency', 'perceived_stress_score', 'self_reported_happiness', 'body_mass_index', 'blood_pressure_systolic', 'blood_pressure_diastolic', 'daily_steps_count', 'weekly_work_hours', 'hobbies_count', 'social_events_per_month', 'books_read_per_year', 'volunteer_hours_per_month', 'travel_frequency_per_year', 'daily_active_minutes_instagram', 'sessions_per_day', 'posts_created_per_week', 'reels_watched_per_day', 'stories_viewed_per_day', 'likes_given_per_day', 'comments_written_per_day', 'dms_sent_per_week', 'dms_received_per_week', 'ads_viewed_per_day', 'ads_clicked_per_day', 'time_on_feed_per_day', 'time_on_explore_per_day', 'time_on_messages_per_day', 'time_on_reels_per_day', 'followers_count', 'following_count', 'uses_premium_features', 'not

,user_id,app_name,age,gender,country,urban_rural,income_level,employment_status,education_level,relationship_status,...,last_login_date,average_session_length_minutes,content_type_preference,preferred_content_theme,privacy_setting_level,two_factor_auth_enabled,biometric_login_used,linked_accounts_count,subscription_status,user_engagement_score
0,1,Instagram,51,Female,India,Rural,High,Retired,Bachelor’s,Single,...,2025-11-02,5.0,Mixed,Tech,Private,Yes,No,0,Free,7.83
1,2,Instagram,64,Female,United Kingdom,Urban,Middle,Full-time employed,Other,Divorced,...,2025-03-22,14.8,Photos,Fashion,Public,No,No,3,Free,1.43
2,3,Instagram,41,Female,Canada,Urban,Middle,Student,Bachelor’s,In a relationship,...,2025-08-10,5.0,Mixed,Other,Public,Yes,Yes,1,Free,9.67
3,4,Instagram,27,Non-binary,South Korea,Urban,Middle,Unemployed,Master’s,In a relationship,...,2025-03-31,25.9,Stories,Tech,Private,No,No,1,Free,0.94
4,5,Instagram,55,Male,India,Urban,Upper-middle,Full-time employed,Bachelor’s,Single,...,2025-03-19,13.1,Videos,Food,Public,Yes,No,0,Free,1.03


In [3]:
print(df.shape)

(1547896, 58)


In [4]:
print(df.dtypes)

user_id                             int64
app_name                           object
age                                 int64
gender                             object
country                            object
urban_rural                        object
income_level                       object
employment_status                  object
education_level                    object
relationship_status                object
has_children                       object
exercise_hours_per_week           float64
sleep_hours_per_night             float64
diet_quality                       object
smoking                            object
alcohol_frequency                  object
perceived_stress_score              int64
self_reported_happiness             int64
body_mass_index                   float64
blood_pressure_systolic             int64
blood_pressure_diastolic            int64
daily_steps_count                   int64
weekly_work_hours                 float64
hobbies_count                     

In [5]:
print(df.isnull().sum())

user_id                           0
app_name                          0
age                               0
gender                            0
country                           0
urban_rural                       0
income_level                      0
employment_status                 0
education_level                   0
relationship_status               0
has_children                      0
exercise_hours_per_week           0
sleep_hours_per_night             0
diet_quality                      0
smoking                           0
alcohol_frequency                 0
perceived_stress_score            0
self_reported_happiness           0
body_mass_index                   0
blood_pressure_systolic           0
blood_pressure_diastolic          0
daily_steps_count                 0
weekly_work_hours                 0
hobbies_count                     0
social_events_per_month           0
books_read_per_year               0
volunteer_hours_per_month         0
travel_frequency_per_year   

In [6]:
print(df.duplicated().sum())

0


In [7]:

print(df['app_name'].unique())

['Instagram']


In [8]:
categorical_cols = df.select_dtypes(include='object').columns.tolist()
print(categorical_cols)
print("Total categorical columns:", len(categorical_cols))

['app_name', 'gender', 'country', 'urban_rural', 'income_level', 'employment_status', 'education_level', 'relationship_status', 'has_children', 'diet_quality', 'smoking', 'alcohol_frequency', 'uses_premium_features', 'last_login_date', 'content_type_preference', 'preferred_content_theme', 'privacy_setting_level', 'two_factor_auth_enabled', 'biometric_login_used', 'subscription_status']
Total categorical columns: 20


In [9]:
# How Much unique values
print(df.nunique())

user_id                           1547896
app_name                                1
age                                    53
gender                                  4
country                                10
urban_rural                             3
income_level                            5
employment_status                       6
education_level                         6
relationship_status                     5
has_children                            2
exercise_hours_per_week               241
sleep_hours_per_night                  71
diet_quality                            5
smoking                                 3
alcohol_frequency                       5
perceived_stress_score                 41
self_reported_happiness                10
body_mass_index                       286
blood_pressure_systolic                70
blood_pressure_diastolic               40
daily_steps_count                     773
weekly_work_hours                     800
hobbies_count                     

## Feature Selection

- `user_id`: Identifier column, dropped (no predictive value)
- `app_name`: Constant column (all values = "Instagram"), dropped
- `last_login_date`: Raw date string, dropped for simplicity
- Remaining categorical columns retained for now, to be encoded in the preprocessing step

# **Drop Irrelevant Columns**

In [10]:
df = df.drop(['user_id', 'app_name', 'last_login_date'], axis=1)

In [11]:
print(df.shape)
# Last Shape was (1547896, 58)

(1547896, 55)


In [12]:
print(df.columns.tolist())

['age', 'gender', 'country', 'urban_rural', 'income_level', 'employment_status', 'education_level', 'relationship_status', 'has_children', 'exercise_hours_per_week', 'sleep_hours_per_night', 'diet_quality', 'smoking', 'alcohol_frequency', 'perceived_stress_score', 'self_reported_happiness', 'body_mass_index', 'blood_pressure_systolic', 'blood_pressure_diastolic', 'daily_steps_count', 'weekly_work_hours', 'hobbies_count', 'social_events_per_month', 'books_read_per_year', 'volunteer_hours_per_month', 'travel_frequency_per_year', 'daily_active_minutes_instagram', 'sessions_per_day', 'posts_created_per_week', 'reels_watched_per_day', 'stories_viewed_per_day', 'likes_given_per_day', 'comments_written_per_day', 'dms_sent_per_week', 'dms_received_per_week', 'ads_viewed_per_day', 'ads_clicked_per_day', 'time_on_feed_per_day', 'time_on_explore_per_day', 'time_on_messages_per_day', 'time_on_reels_per_day', 'followers_count', 'following_count', 'uses_premium_features', 'notification_response_rate

In [13]:
print(df.shape)

(1547896, 55)


# **Step: One-Hot Encoding**

In [14]:
categorical_cols = df.select_dtypes(include='object').columns.tolist()
print(categorical_cols)

['gender', 'country', 'urban_rural', 'income_level', 'employment_status', 'education_level', 'relationship_status', 'has_children', 'diet_quality', 'smoking', 'alcohol_frequency', 'uses_premium_features', 'content_type_preference', 'preferred_content_theme', 'privacy_setting_level', 'two_factor_auth_enabled', 'biometric_login_used', 'subscription_status']


# **One-Hot Encoding Apply**

In [15]:
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

In [16]:
print("Shape after encoding:", df_encoded.shape)

Shape after encoding: (1547896, 99)


In [17]:
df_encoded.head()

,age,exercise_hours_per_week,sleep_hours_per_night,perceived_stress_score,self_reported_happiness,body_mass_index,blood_pressure_systolic,blood_pressure_diastolic,daily_steps_count,weekly_work_hours,...,preferred_content_theme_Music,preferred_content_theme_Other,preferred_content_theme_Tech,preferred_content_theme_Travel,privacy_setting_level_Private,privacy_setting_level_Public,two_factor_auth_enabled_Yes,biometric_login_used_Yes,subscription_status_Free,subscription_status_Premium
0,51,7.2,7.7,3,8,20.8,148,86,8107,49.9,...,False,False,True,False,True,False,True,False,True,False
1,64,10.9,8.6,1,1,23.5,133,84,8059,15.6,...,False,False,False,False,False,True,False,False,True,False
2,41,5.0,6.7,4,10,28.6,135,88,7872,31.8,...,False,True,False,False,False,True,True,True,True,False
3,27,10.6,6.5,18,1,22.5,105,73,7801,43.4,...,False,False,True,False,True,False,False,False,True,False
4,55,7.7,6.8,19,1,28.1,146,90,8005,50.2,...,False,False,False,False,False,True,True,False,True,False


In [18]:
print("df shape (before encoding):", df.shape)
print("df_encoded shape (after encoding):", df_encoded.shape)

df shape (before encoding): (1547896, 55)
df_encoded shape (after encoding): (1547896, 99)


# **Train-Test Split**

In [19]:
from sklearn.model_selection import train_test_split

In [20]:
X = df_encoded.drop('user_engagement_score', axis=1)

In [21]:
y = df_encoded['user_engagement_score']

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [23]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (1238316, 98)
X_test shape: (309580, 98)


In [24]:
from sklearn.linear_model import LinearRegression

In [25]:
model = LinearRegression()

In [26]:
model.fit(X_train, y_train)

LinearRegression()

In [27]:
LinearRegression()

LinearRegression()

In [28]:
print(X_train.shape)

(1238316, 98)


# **Model successfully train**

# **Next Step: Model Prediction**

In [29]:
y_pred = model.predict(X_test)

In [30]:
y_pred[:10]

array([1.95022925, 1.23122701, 1.62030898, 4.42449895, 1.09393357,
       2.26599252, 0.72548923, 1.35880403, 0.92849232, 0.4923678 ])

# **Model Evaluation**
**Calculate Metrics**

In [31]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [32]:
mae = mean_absolute_error(y_test, y_pred)

In [33]:
mse = mean_squared_error(y_test, y_pred)

In [34]:
rmse = np.sqrt(mse)

In [35]:
r2 = r2_score(y_test, y_pred)

In [36]:
print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R² Score:", r2)

MAE: 0.8288311791448362
MSE: 1.4600063255985514
RMSE: 1.208307214907927
R² Score: 0.5567372373635813


# **Quick Interpretation**

In [37]:
print(df['user_engagement_score'].describe())

count    1.547896e+06
mean     1.644642e+00
std      1.812684e+00
min      6.700000e-01
25%      1.000000e+00
50%      1.100000e+00
75%      1.290000e+00
max      1.867000e+01
Name: user_engagement_score, dtype: float64


#  ** Model Performance Analysis**

The Linear Regression model achieved an R² Score of 0.557, meaning it explains about 56% of 
the variance in user engagement scores. This indicates moderate predictive performance.

The MAE of 0.83 is significant relative to the target variable's mean (1.64) and median (1.10) — 
this means predictions are off by roughly 50-75% of a typical engagement score, which is a 
considerable margin of error for practical use.

The target variable (user_engagement_score) is highly right-skewed (min: 0.67, max: 18.67, 
75th percentile: only 1.29), meaning most users have low engagement scores while a small number 
of users have very high scores. Linear Regression assumes a roughly linear, normally-distributed 
relationship, so this skewness likely limits model accuracy.

**Conclusion:** The model shows moderate performance. Potential improvements include:
- Applying a log transformation to the target variable to reduce skewness
- Trying non-linear models (e.g., Random Forest, Gradient Boosting) which handle skewed data better
- Investigating which features have the strongest correlation with engagement score